library/network/src/modules/SuSEFirewallServices.rb

```diff
@@ -61,6 +54,175 @@ class SuSEFirewallServicesClass < Module
+    end
+
+    # Function returns needed IP protocols for service
+    #
+    # @param [String] service
+    # @return	[Array<String>] of needed IP protocols
+    def GetNeededIPProtocols(service)
+      service_details(service)["ip_protocols"] || []
+    end
+
+    # Function returns description of a firewall service
+    #
+    # @param [String] service
+    # @return	[String] service description
+    def GetDescription(service)
+      service_details(service)["description"] || []
+    end
+
+    # Function returns needed ports and protocols for service.
+    # Service needs to be known (installed in the system).
+    # Function throws an exception SuSEFirewalServiceNotFound
+    # if service is not known (undefined).
+    #
+    # @param [String] service
+    # @return	[Hash{String => Array<String>}] of needed ports and protocols
+    #
+    # @example
+    #	GetNeededPortsAndProtocols ("service:aaa") -> $[
+    #		"tcp_ports"       : [ "122", "ftp-data" ],
+    #		"udp_ports"       : [ "427" ],
+    #		"rpc_ports"       : [ "portmap", "ypbind" ],
+    #		"ip_protocols"    : [],
+    #		"broadcast_ports" : [ "427" ],
+    #	];
+    def GetNeededPortsAndProtocols(service)
+      DEFAULT_SERVICE.merge(service_details(service))
+    end
+
+    # Returns all known services loaded from disk on-the-fly
+    def all_services
+      ReadServicesDefinedByRPMPackages() if @services.nil?
+      @services
+    end
+
+    # Sets that configuration was not modified
+    def ResetModified
+      @sfws_modified = false
+
+      nil
+    end
+
+    # Returns whether configuration was modified
+    #
+    # @return [Boolean] modified
+    def GetModified
+      @sfws_modified
+    end
+
+    # Create appropriate firewall services class based on factors such as which
+    # backend is selected by user or running on the system.
+    #
+    # @note If the backend_sym parameter is :fwd (ie, FirewallD is the desired
+    # @note firewall backend), then the method will also start the FirewallD service.
+    #
+    # @param backend_sym [Symbol] if not nil, explicitely select :sf2 or :fwd
+    # @return SuSEFirewall2ServicesClass or SuSEfirewalldServicesClass instance
+    def self.create(backend_sym = nil)
+      # If backend is specificed, go ahead and create an instance. Otherwise, try
+      # to detect which backend is enabled and create the appropriate instance.
+      if backend_sym == :sf2
+        SuSEFirewall2ServicesClass.new
+      elsif backend_sym == :fwd
+        # We need to start the backend first
+        if !SuSEFirewall.IsStarted()
+          log.info "Starting the FirewallD service"
+          SuSEFirewall.StartServices()
+        end
+        SuSEFirewalldServicesClass.new
+      else
+        # Instantiate one based on the running backend
+        if SuSEFirewall.is_a?(SuSEFirewall2Class)
+          SuSEFirewall2ServicesClass.new
+        else
+          SuSEFirewalldServicesClass.new
+        end
+      end
+    end
```

Comment:
well, it do not handle other symbols and simple ignore it, which is error prone. Also code can benefit from case method so it can look like
```
  case backend_sym
  when :sf2
     SuSEFirewall2ServicesClass.new
  when :fwd
     # We need to start the backend first
     if !SuSEFirewall.IsStarted()
       log.info "Starting the FirewallD service"
       SuSEFirewall.StartServices()
    end
    SuSEFirewalldServicesClass.new
when nil
  # Instantiate one based on the running backend
  if SuSEFirewall.is_a?(SuSEFirewall2Class)
    SuSEFirewall2ServicesClass.new
  else
     SuSEFirewalldServicesClass.new
  end
else
  raise "Invalid symbol for firewall backend #{backend_sym.inspect}"
end
```